### PR TITLE
Start to generate the CLI options dynamically

### PR DESCRIPTION
**Pull request by wilzbach**
Problem:
- https://dlang.org/dmd.html (there are four pages) are nearly __always__ out of sync
- The MAN pages are nearly __always__ out of sync

Solution: Have one source of truth and generate MAN pages + the dlang.org documentation.
This PR just moves the available CLI flags into a separate module + makes them available as an array for consumption.

As I didn't know whether it would be ok to simply loop over them and use `printf` at runtime, I generated one big string in CTFE, s.t. hopefully this isn't is blocking this first step.

Future steps:
- Make `mpcu` and `transitionUsage` `Options` too
- Use this file to generate the documentation on dlang.org
- Use this file to generate the MAN pages
- Combine this file with `globals.d` and automatically generate `parseCommandLine`

### CODE REVIEW

**Comment by JinShil** [comment]
> Make mpcu and transitionUsage Options too
> Combine this file with globals.d and automatically generate parseCommandLine

I've had a couple of bad experiences already approving and merging partial solutions.  Can include the above two items into this pull request.

> Use this file to generate the MAN pages

How are the man pages currently generated?

**Comment by wilzbach** [comment]
> I've had a couple of bad experiences already approving and merging partial solutions. Can you include the above two items into this pull request?

And I made bad experiences with doing too much work in one PR ;-)
I purposefully didn't do more work in this PR, because I wasn't sure whether the way I did it, was ok.

Anyhow, it's not a partial solution, it's an incremental improvement.
With this PR approved/merged. I can already generate the MAN page (we don't list the `transitions` nor `mcpu` there atm). 

> How are the man pages currently generated?

They aren't really _generated_:

https://github.com/dlang/dmd/blob/master/docs/man/man1/dmd.1

They are static files which are rarely updated, because people who add new flags forget about them (to be fair - it's very easy to do so).

**Comment by JinShil** [comment]
Does `cli.d` need to also be added to the `src/vcbuild/dmd.visualdproj` file?

**Comment by wilzbach** [comment]
Added a script that generated the MAN page based on the options array in `cli.d`.
-> I pushed the "updated" version of the MAN file in a separate commit
It comes with a simple Makefile that can be used to `build` or `preview` the MAN page. It does no work if `cli.d` nor `gen_man.d` haven't changed, so it will be easy to integrate it with the main Makefile in a later PR.
I hope the vast differences between the MAN page and actual flags show why we want to automate this.

**Comment by wilzbach** [comment]
Anything that's blocking this? I'm getting tired of rebasing this whenever the CLI options change (and we should have updated the MAN page + dlang.org docs too).
Also I just bumped into three bugzilla issues about outdated MAN pages:
- https://issues.dlang.org/show_bug.cgi?id=14475
- https://issues.dlang.org/show_bug.cgi?id=15679
- https://issues.dlang.org/show_bug.cgi?id=15446

This PR is already complete in itself and a good step forward (the MAN pages are generated).
- Dlang.org's docs are more complicated -> I will submit this in a separate PR
- Automatically rebuilding the MAN page as part of the build might be controversial -> I will submit this in a separate PR

File: docs/Makefile
Comment (wilzbach) [inline docs/Makefile]: The main mainfile can call this target to regenerate the man page when `cli.d` changed, but as I said before, I think the discussion whether that's a good idea or not, shouldn't block this PR and thus belongs into another PR.

**Comment by wilzbach** [comment]
There's more:
- https://issues.dlang.org/show_bug.cgi?id=11985
- https://issues.dlang.org/show_bug.cgi?id=8196

File: src/dmd/cli.d
Comment (JinShil) [inline src/dmd/cli.d]: Can we return a `string` and use `.ptr` at the `printf` call.  I'd like to gradually move away from using `char*`.

**Review by JinShil — APPROVED** [review]

File: src/dmd/cli.d
Comment (jacob-carlborg) [inline src/dmd/cli.d]: Looks like this can be a variable instead of a function.

File: src/dmd/cli.d
Comment (jacob-carlborg) [inline src/dmd/cli.d]: Opening curly brace on the next line.

**Comment by jacob-carlborg** [comment]
Some of the comments are Ddoc comments and some are regular comments. Is there a reason for that?

File: src/dmd/cli.d
Comment (jacob-carlborg) [inline src/dmd/cli.d]: Looks like this can be private.

File: src/dmd/cli.d
Comment (jacob-carlborg) [inline src/dmd/cli.d]: Remove extra newline.

File: src/dmd/cli.d
Comment (jacob-carlborg) [inline src/dmd/cli.d]: Missing documentation for these three fields.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

File: src/dmd/cli.d
Comment (jacob-carlborg) [reply src/dmd/cli.d]: Or even better not use `.ptr` at all and instead use the `printf` formatter that allows to specify pointer and length separately.

Comment (wilzbach) [reply src/dmd/cli.d]: Me too, I just used `char*` to keep the changes at a minimum. I'm more than happy to use `string` :)

**Comment by wilzbach** [comment]
> Some of the comments are Ddoc comments and some are regular comments. Is there a reason for that? > Otherwise all public symbols should have Ddoc comments.

Oh I wanted to leave everything undocumented for now as this might still change when I integrate it with dlang.org.
(I removed the Ddoc star for the symbol in question).

OTOH we don't give any guarantees about the API stability at the moment, so we could also simply simple start documenting it. (I don't have a strong opinion here)

File: src/dmd/mars.d
Comment (jacob-carlborg) [inline src/dmd/mars.d]: Is there a reason why `&help[0]` is used instead of `help.ptr`?. `.ptr` is more readable in my opinion.

**Comment by jacob-carlborg** [comment]
> OTOH we don't give any guarantees about the API stability at the moment, so we could also simply simple start documenting it. (I don't have a strong opinion here)

I'm in favor of documenting it.

File: src/dmd/mars.d
Comment (wilzbach) [reply src/dmd/mars.d]: Yes `help.ptr` isn't `@safe`

**Comment by wilzbach** [comment]
> I'm in favor of documenting it.

Fair enough. Done.

File: src/dmd/mars.d
Comment (jacob-carlborg) [reply src/dmd/mars.d]: Eh, ok. I didn't know there was a difference.

File: src/dmd/cli.d
Comment (jacob-carlborg) [inline src/dmd/cli.d]: Missing documentation.

File: src/dmd/cli.d
Comment (jacob-carlborg) [inline src/dmd/cli.d]: Not a Ddco comment.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

**Comment by jacob-carlborg** [comment]
I would prefer that the existing commits are amended, then we avoid unnecessary commits like `Address review`. Unless you want to squash the whole thing.

**Comment by wilzbach** [comment]
> I would prefer that the existing commits are amended

Sure. Done. Just did individual commits, s.t. it's easier to follow up.

**Comment by jacob-carlborg** [comment]
> Sure. Done. Just did individual commits, s.t. it's easier to follow up.

You could do fixup commits in that case.

**Comment by jacob-carlborg** [comment]
LGTM 👌.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
Probably someone else should have a look as well.

**Comment by wilzbach** [comment]
> You could do fixup commits in that case.

- Start to generate the CLI options dynamically
<- I wanted to squashed things here
- Add script to automatically generate the MAN pages
- Use the generated MAN pages

**Comment by jacob-carlborg** [comment]
Fixup commits allow you to fixup any commit and not just the last one.

**Review by JinShil — APPROVED** [review]

**Review by PetarKirov — APPROVED** [review]
Overall looking great! As discussed with Seb, there are various parts that could be improved, but it's much better do move forward with this and improve it over time.

**Pull request merged**
